sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteRowLevelCommand.scala

```diff
@@ -132,52 +164,119 @@ trait RewriteRowLevelCommand extends Rule[LogicalPlan] {
+      val outputsWithRow = filterOutputs(outputs, DELTA_OPERATIONS_WITH_ROW)
+      Some(newLazyProjection(plan, outputsWithRow, rowAttrs))
     } else {
       None
     }
 
-    val rowIdProjection = newLazyRowIdProjection(plan, rowIdAttrs)
+    val outputsWithRowId = filterOutputs(outputs, DELTA_OPERATIONS_WITH_ROW_ID)
+    val rowIdProjection = newLazyRowIdProjection(plan, outputsWithRowId, rowIdAttrs)
 
     val metadataProjection = if (metadataAttrs.nonEmpty) {
-      Some(newLazyProjection(plan, metadataAttrs))
+      val outputsWithMetadata = filterOutputs(outputs, DELTA_OPERATIONS_WITH_METADATA)
+      Some(newLazyProjection(plan, outputsWithMetadata, metadataAttrs))
     } else {
       None
     }
 
     WriteDeltaProjections(rowProjection, rowIdProjection, metadataProjection)
   }
 
+  private def extractOutputs(plan: LogicalPlan): Seq[Seq[Expression]] = {
+    plan match {
+      case p: Project => Seq(p.projectList)
+      case e: Expand => e.projections
+      case m: MergeRows => m.outputs
+      case _ => throw SparkException.internalError("Can't extract outputs from plan: " + plan)
+    }
+  }
+
+  private def filterOutputs(
+      outputs: Seq[Seq[Expression]],
+      operations: Set[Int]): Seq[Seq[Expression]] = {
+    outputs.filter {
+      case Literal(operation: Integer, _) +: _ => operations.contains(operation)
+      case Alias(Literal(operation: Integer, _), _) +: _ => operations.contains(operation)
+      case other => throw SparkException.internalError("Can't determine operation: " + other)
+    }
+  }
+
   private def newLazyProjection(
       plan: LogicalPlan,
+      outputs: Seq[Seq[Expression]],
       attrs: Seq[Attribute]): ProjectingInternalRow = {
-
     val colOrdinals = attrs.map(attr => findColOrdinal(plan, attr.name))
-    val schema = DataTypeUtils.fromAttributes(attrs)
-    ProjectingInternalRow(schema, colOrdinals)
+    createProjectingInternalRow(outputs, colOrdinals, attrs)
   }
 
   // if there are assignment to row ID attributes, original values are projected as special columns
   // this method honors such special columns if present
   private def newLazyRowIdProjection(
       plan: LogicalPlan,
+      outputs: Seq[Seq[Expression]],
       rowIdAttrs: Seq[Attribute]): ProjectingInternalRow = {
-
     val colOrdinals = rowIdAttrs.map { attr =>
       val originalValueIndex = findColOrdinal(plan, ORIGINAL_ROW_ID_VALUE_PREFIX + attr.name)
       if (originalValueIndex != -1) originalValueIndex else findColOrdinal(plan, attr.name)
     }
-    val schema = DataTypeUtils.fromAttributes(rowIdAttrs)
+    createProjectingInternalRow(outputs, colOrdinals, rowIdAttrs)
+  }
+
+  private def createProjectingInternalRow(
+      outputs: Seq[Seq[Expression]],
+      colOrdinals: Seq[Int],
+      attrs: Seq[Attribute]): ProjectingInternalRow = {
+    val schema = StructType(attrs.zipWithIndex.map { case (attr, index) =>
+      val nullable = outputs.exists(output => output(colOrdinals(index)).nullable)
```

Comment:
@aokolnychyi Do we only need this for metadata columns? For regular columns, shall we use `attr.nullable` instead?